Adds two more Learn PWA modules (#6926)
GoogleChrome/web.dev: src/site/content/en/learn/pwa/caching/index.md

---
title: Caching
description: >

>   With the Cache Storage API, you can ask the browser to download, store, delete or update assets on the device. The engine can then serve these assets on the device without needing a network request.

Reviewer: I'm hesitant of removing the "you can ask" because I still want to convey that the developers have agency with the API.

Maybe rewriting it as:
 "You can use the Cache Storage API to download, store, delete or update assets on the device. Then these assets can be served on the device without needing a network request."

or since this is the intro now that I think about it I don't know why it only speaks about the cache storage api, maybe:
"Caches are powerful tools that enable fast loading and offline experiences on the web. In this chapter you'll learn best practices and how to use the he Cache Storage API to download, store, delete or update assets on the device. Then serve those assets  without needing a network request."

>   With the Cache Storage API, you can ask the browser to download, store, delete or update assets on the device. The engine can then serve these assets on the device without needing a network request.

Reviewer 2: I prefer your first suggestion. 

So many writers describe APIs as 'powerful' that the word has lost its meaning smacks of hyperbole.

authors:
  - firt
date: 2021-12-03
---

> Cache storage is a powerful tool. It makes our apps less dependent on network conditions. With good use of caches you can make your web app available offline and serve your assets as fast as possible in any network condition As mentioned in [Assets and Data](/learn/pwa/assets-and-data/) you can decide the best strategy for caching the necessary assets. To manage the cache your service worker can interact with the [Cache Storage API](https://developer.mozilla.org/docs/Web/API/CacheStorage).

Reviewer: ```suggestion
Cache storage is a powerful tool. It makes our apps less dependent on network conditions. With good use of caches you can make your web app available offline and serve your assets as fast as possible in any network condition. As mentioned in [Assets and Data](/learn/pwa/assets-and-data/) you can decide the best strategy for caching the necessary assets. To manage the cache your service worker can interact with the [Cache Storage API](https://developer.mozilla.org/docs/Web/API/CacheStorage).
```

{% Aside %}
When installing platform-specific apps, the device stores the icon and other app assets in the operating system, in one step. For PWAs, the process has two separate steps. A PWA  can store assets on your device any time after the first visit in the browser, even without installation. Install is a separate action that is covered later in this course.
{% endAside %}

The Cache Storage API is available from different contexts: 

- The window context (your PWA's main thread).
- The service worker.
- Any other workers you use. 

One advantage of managing your cache using service workers is that its lifecycle is not tied to the window, which means you are not blocking the main thread. But be aware that to use the Cache Storage API most of these contexts have to be under a TLS connection.

## What to cache

The first question you may have when you know you can cache or install assets on a user's device for later usage is what to cache. While there is no single answer to that question, you can start with all the minimum resources that you need to render the user interface.

That should include:

- The main entry point HTML (your app's start_url).

> - Key CSS stylesheets.

Author: Maybe but I think it is a bit vague and app dependent. Above it says that you should cache the minimum resources to load the UI, and I think that's probably what we are driving at.

> - Key CSS stylesheets.

Author: ```suggestion
- CSS stylesheets needed for the main user interface.
```

> - Images from the user interface.

Reviewer: ```suggestion
- Images from used in user interface.
```

> - JavaScript files you need to render the user interface.

Reviewer: ```suggestion
- JavaScript files required to render the user interface.
```

> - Data you need to render a basic experience, such as a JSON file.

Reviewer: ```suggestion
- Data required to render a basic experience, such as a JSON file.
```

{% endAside %}

The Cache Storage API is available from different contexts:

- The window context (your PWA's main thread).
- The service worker.
- Any other workers you use.

One advantage of managing your cache using service workers is that its lifecycle is not tied to the window, which means you are not blocking the main thread. Be aware that to use the Cache Storage API most of these contexts have to be under a TLS connection.

## What to cache

The first question you may have about caching is what to cache. While there is no single answer to that question, you can start with all the minimum resources that you need to render the user interface.

Those resources should include:

- The main page HTML (your app's start_url).
- Key CSS stylesheets.
- Images used in the user interface.
- JavaScript files required to render the user interface.

> - Data, such as a JSON file, that you need to render a basic experience.

Author: ```suggestion
- Data, such as a JSON file, required to render a basic experience.
```
